### PR TITLE
Graceful reconnection alternative

### PR DESCRIPTION
**Pull request by javiercn**
* Uses the send beacon API instead of relying on SignalR disconnect feature
* Works as we expect in the three cases we care about:
  * User refreshes the window.
  * User closes the browser.
  * User navigates away from the page.
* It introduces an additional POST endpoint at path+/disconnect that receives the circuit id in a form post value and processes it.
  * We don't need to worry about CSRF here because the circuit ID is not guessable.
  * If we want to be extra careful we could also require the circuit being disconnected from the hub before we fully terminate it. (Meaning the beacon would simply add a flag to the circuit and Registry.DisconnectAsync would check this flag to take down the circuit completely or we would take down the circuit completely if it was already disconnected when we received the flag).
* None of these things are in reality a problem as DisconnectAsync already has logic to handle these and the beacon gets sent when the page unloads, so there is little/no chance of receiving any interop/event call at that point.

image

### CODE REVIEW

**Comment by rynowak** [comment]
@GrabYourPitchforks FYI

I want to make sure we get a signoff re: security for something we're adding this late in the game.

I'm impressed on the quick turnaround on this 😆

File: src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs
Comment (rynowak) [inline src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs]: fix spacing :)

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (rynowak) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]: Is the reason to use a middleware here just so you can use DI? If so, consider adding a comment.

File: src/Components/Server/src/Builder/ComponentEndpointRouteBuilderExtensions.cs
Comment (rynowak) [inline src/Components/Server/src/Builder/ComponentEndpointRouteBuilderExtensions.cs]: Solid 👍 

Routerware is the best thing ever.

File: src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs
Comment (rynowak) [inline src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs]: any real reason for params here? You always pass two.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (rynowak) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]: it would be a good idea to validate the content type here or else this will throw.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (rynowak) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]: nit: spacing

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (rynowak) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]: nit: 400 is probably a better status code.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (rynowak) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]: Should this be after the call to `Terminate`? The log is going to say that we terminated gracefully, even if it throws.

File: src/Components/Server/src/Circuits/CircuitRegistry.cs
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitRegistry.cs]: Is this the only caller of `PermanentDisconnect`? consider just inlining that code here. I'm not sure we want other callers to separate the logic of `PermanentDisconnect` from the dispose logic. 

I'll end up calling this code later in my next work item, and I'd prefer to call `Terminate`.

File: src/Components/Server/src/Circuits/CircuitRegistry.cs
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitRegistry.cs]: nit: public. This should probably be `TerminateAsync` for consistency with other things.

File: src/Components/Server/src/Circuits/CircuitRegistry.cs
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitRegistry.cs]: This logic (line 309-321) seems like it could be simplified to a single `if (circuitHost != null)` or `circuitHost?.DisposeAsync() ?? default`.

File: src/Components/Server/test/ComponentEndpointRouteBuilderExtensionsTest.cs
Comment (rynowak) [inline src/Components/Server/test/ComponentEndpointRouteBuilderExtensionsTest.cs]: What changed that made this necessary?

**Review by rynowak — COMMENTED** [review]
Some minor feedback, this looks really good so far.

**Comment by rynowak** [comment]
Tests? Both for the new endpoint and for the disconnection functionality in general? We might need to add a "test" endpoint to the server that can be used to query circuits.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (SteveSandersonMS) [reply src/Components/Server/src/CircuitDisconnectMiddleware.cs]: Also, the use of form-encoded data here is quite unusual, but I like it. Simpler than having to have a type to JSON-deserialize to 👍

File: src/Components/Web.JS/src/Boot.Server.ts
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Boot.Server.ts]: The `_` param is strange. Does the compiler complain if you just make it `sendDisconnect()`? TBH it might be nicer to inline this into line 50 anyway.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (campersau) [inline src/Components/Server/src/CircuitDisconnectMiddleware.cs]:
```suggestion
```

File: src/Components/Web.JS/src/Boot.Server.ts
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Boot.Server.ts]: Is it technically necessary to supply an absolute URL for `sendBeacon`? Does this fail if it was just:

```ts
navigator.sendBeacon(`_blazor/disconnect`, data);
```

I can't see anything in the spec that seems to imply this wouldn't work. The example [here](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon) uses a relative URL.

**Review by SteveSandersonMS — APPROVED** [review]
I love this! Seems so much clearer than trying to guess based on heartbeats.

I'm approving now because this looks so good, but I'm sure you'll address the comments too.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (javiercn) [reply src/Components/Server/src/CircuitDisconnectMiddleware.cs]: We send a FormData object on the client. The send beacon API doesn't give you control over the request, (AFAIK). We could have passed a string, but that likely arrived as text/plain.

I think using FormData is a better choice for this particular case. I'll add a check here for the content type.

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (javiercn) [reply src/Components/Server/src/CircuitDisconnectMiddleware.cs]: Fair, will move.

File: src/Components/Web.JS/src/Boot.Server.ts
Comment (javiercn) [reply src/Components/Web.JS/src/Boot.Server.ts]: Fair enough, I just put this together before doing cleanup.

**Comment by SteveSandersonMS** [comment]
Based on @anurse's comments elsewhere, maybe at some point in the future SignalR could introduce a new method for notifying the app about *definitely-graceful* disconnections, if it turns out we can reliably distinguish between things like "user switched phone into flight mode" (ungraceful) and "use navigated away from page" (graceful).

If that does become true in the future, maybe that would be even more reliable than using beacons, since it would be harder for a misbehaving client to mislead us. We'd be free to switch to a different implementation, since this beacon technique is purely an internal implementation detail. For now, the beacon definitely looks like the best bet to me.

File: src/Components/Web.JS/src/Boot.Server.ts
Comment (campersau) [inline src/Components/Web.JS/src/Boot.Server.ts]: Not sure what browser versions you want to support, because `set` is actually quite _new_ and `append` would also work in this case.
https://developer.mozilla.org/en-US/docs/Web/API/FormData/append#Browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/API/FormData/set#Browser_compatibility

(Mabye also add a check for `sendBeacon`?)

Comment (javiercn) [reply src/Components/Web.JS/src/Boot.Server.ts]: set is supported on all browsers we care about. I will change it to append though because its trivial.

sendBeacon is also supported on all browsers we care about, and even if that's not the case you would get an error when you are unloading the document, so that's not a problem.

File: src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs
Comment (javiercn) [reply src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs]: done

File: src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs
Comment (javiercn) [reply src/Components/Server/src/Builder/ComponentEndpointConventionBuilder.cs]: I was just following the pattern signalr follows. It feels pretty straightforward and its all internal. Do you feel strongly against it? I've changed it, but I don't think it's important at all

File: src/Components/Server/src/CircuitDisconnectMiddleware.cs
Comment (javiercn) [reply src/Components/Server/src/CircuitDisconnectMiddleware.cs]: Is it though?

I used it for DI, but also because its better to have a middle-ware you can tests than a bunch of code hidden within the extension method.

File: src/Components/Server/test/ComponentEndpointRouteBuilderExtensionsTest.cs
Comment (javiercn) [reply src/Components/Server/test/ComponentEndpointRouteBuilderExtensionsTest.cs]: I think the routerware

**Comment by javiercn** [comment]
:up: :date: Will merge tomorrow unless there's important feedback

**Pull request merged**
